### DISCUSSION TITLE
Validity of Action preconditions and effects

### DISCUSSION THREAD

**Discussion by e-cal** [AI-Planning/macq#61] [General]
In the original base API we implemented a method to check the validity of a fluent being added as a precondition or effect. The method checks that the `Fluent` and `Action` share at least one object, making them relevant to eachother. In implementing Observer, I came across an interesting problem where seemingly irrelevant fluents can be classified as preconditions (or effects) despite acting on different objects, which raises the exception in the validity check.

My example is a simple blocks world with 2 blocks, and the issue arises with the action "Pick up A" and the precondition "B on table". The fluent "B on table" is always true in the pre-state to "Pick up A", so it is a precondition according to observer, but because it doesn't include the object A it is invalid.

I think the solution would be to remove the validity check from `Action` and leave it all up to the extraction technique.
Or if it is important that irrelevant fluents aren't added to an action, to simply not add the fluent and print a warning instead of raising an exception.

**Comment by e-cal** [comment]
For now I have it printing a warning and add the fluent.

**Reply by haz** [reply]
Hrmz...there are a few things to note with respect to this:

- In the extraction, we may not know what the parameters are until all is said and done
- There may not be _any_ parameters for actions or fluents at times
- There is a difference between constants and objects. The former allows models to use those objects without a parameter. `table` (if there's only one) is typically used as a constant.

So for these reasons, we may want to just forgo trying to be smart about checking validity. As you mention, certain extraction techniques may want to tackle this directly, but that would be situation-specific.
